cardea-server/src/main/java/ca/on/oicr/gsi/cardea/server/service/CaseService.java

```diff
@@ -133,6 +138,72 @@ public Set<ShesmuCase> getShesmuCases() {
         .collect(Collectors.toSet());
   }
 
+  public Set<ShesmuDetailedCase> getShesmuDetailedCases() {
+    return caseData.getCases().stream()
+            .map(kase -> convertCaseToShesmuDetailedCase(kase))
+            .collect(Collectors.toSet());
+  }
+
+  private Set<ShesmuTest> getTestsForShesmuWithStatus(Case kase){
+
+    Set<ShesmuTest> sTests = new HashSet<>();
+    final var reqId = kase.getRequisition().getId();
+
+    for (Test test : kase.getTests()) {
+
+      var lqs = new HashSet<ShesmuSample>();
+      var fdls = new HashSet<ShesmuSample>();
+
+      Boolean hasLQPassed = false;
+      Boolean hasLQWaiting = false;
+      for (Sample lq : test.getLibraryQualifications()){
+        if (lq.getDataReviewPassed() == null || lq.getQcPassed() == null){
+          hasLQWaiting = true;
+        } else if (lq.getDataReviewPassed() && lq.getQcPassed()) {
+          hasLQPassed = true;
+        }
+        if (lq.getRun() != null) {
+          lqs.add(new ShesmuSample.Builder()
+                  .id(lq.getId())
+                  .isSupplemental(!Objects.equals(lq.getRequisitionId(), reqId))
+                  .build());
+        }
+      }
+
+      Boolean hasFDPassed = false;
+      Boolean hasFDWaiting = false;
+      for (Sample fd : test.getFullDepthSequencings()){
+
+        if (fd.getDataReviewPassed() == null || fd.getQcPassed() == null){
+          hasFDWaiting = true;
+        } else if (fd.getDataReviewPassed() && fd.getQcPassed()) {
+          hasFDPassed = true;
+        }
+
+        fdls.add(new ShesmuSample.Builder()
+                .id(fd.getId())
+                .isSupplemental(!Objects.equals(fd.getRequisitionId(), reqId))
+                .build());
+      }
+
+      sTests.add(new ShesmuTest.Builder()
+              .name(test.getName())
+              .limsIds(lqs)
+              .isComplete((hasLQPassed && !hasLQWaiting))
+              .test(TestCategory.LIBRARYQUALIFICATION)
+              .build());
+
+      sTests.add(new ShesmuTest.Builder()
+              .name(test.getName())
+              .limsIds(fdls)
+              .isComplete((hasFDPassed && !hasFDWaiting))
+              .test(TestCategory.FULLDEPTHSEQUENCING)
+              .build());
+    };
```

Comment:
Lots of repetition. Could be reduced to something like

```
for each test
  sTests.add(makeShesmuTest(test.getLibraryQualifications(), LIBRARY_QUALIFICATION, reqId);
  sTests.add(makeShesmuTest(test.getFullDepthSequencings(), FULL_DEPTH_SEQUENCING, reqId);

...

makeShesmuTest(samples, type, caseRequisitionId) {
 ...
}
```

Should there be a `ShesmuTest` at all if there are no run-libraries? What if library qualification uses library aliquots only - no sequencing?